File: Tests/RangerTests.cs
using System;
using System.Text.Patterns;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests {
	[TestClass]
	public class RangerTests {
		[TestMethod]
		public void Constructor() {
			Pattern _ = Pattern.Range(From: "Hello", To: ';');
		}

		[TestMethod]
		public void Consume() {
			Pattern Ranger = Pattern.Range(From: "Hello", To: ';');
			Result Result;

			Result = Ranger.Consume("Hello;");
			ResultAssert.Captures("Hello;", Result);


			Result = Ranger.Consume("Hello World;");
			ResultAssert.Captures("Hello World;", Result);
		}

		[TestMethod]
		public void EscapedConsume() {
			Pattern Ranger = Pattern.Range(From: "\"", To: "\"", Escape: "\"\"");
			Result Result;

			Result = Ranger.Consume("\"\"");
			ResultAssert.Captures("\"\"", Result);

			Result = Ranger.Consume("\"H\"");
			ResultAssert.Captures("\"H\"", Result);

			Result = Ranger.Consume("\"Hello\"");
			ResultAssert.Captures("\"Hello\"", Result);

			Result = Ranger.Consume("\"Hello\"\"Goodbye\"");
			ResultAssert.Captures("\"Hello\"\"Goodbye\"", Result);

			// Should also check for C style escapes
			Ranger = Pattern.Range(From: "\"", To: "\"", Escape: "\\\"");

			Result = Ranger.Consume("\"Hello\\\"Goodbye\"");
			ResultAssert.Captures("\"Hello\\\"Goodbye\"", Result);
		}

		[TestMethod]
		public void NestedConsume() {
			Pattern Ranger = Pattern.NestedRange(From: "if", To: "end if");
			Result Result;

			Result = Ranger.Consume("if\nif\nend if\nbacon\nend if\nfoobar");
			ResultAssert.Captures("if\nif\nend if\nbacon\nend if", Result);

			Result = Ranger.Consume("if\nif\nend if\nbacon\nfoobar");
			ResultAssert.Fails(Result);
		}
	}
}

File: Tests/Patterns/RangerTests.fs
namespace Tests.Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type RangerTests() =

    [<TestMethod>]
    member _.``range consume`` () =
        let pattern = range "Hello" ';'
        let mutable result = Result()

        result <- pattern.Consume("Hello;")
        ResultAssert.Captures("Hello;", result)

        result <- pattern.Consume("Hello World;")
        ResultAssert.Captures("Hello World;", result)

    [<TestMethod>]
    member _.``range neglect`` () =
        let pattern () = negate(range "Hello" ';') |> ignore
        Assert.ThrowsException<PatternConstructionException>(Action pattern) |> ignore

    [<TestMethod>]
    member _.``erange consume``() =
        let pattern = erange "\"" "\"" "\\\""
        let mutable result = Result()

        result <- pattern.Consume("\"\"")
        ResultAssert.Captures("\"\"", result)

        result <- pattern.Consume("\"H\"")
        ResultAssert.Captures("\"H\"", result)

        result <- pattern.Consume("\"Hello\"")
        ResultAssert.Captures("\"Hello\"", result)

        result <- pattern.Consume("\"Hello\\\"Goodbye\"")
        ResultAssert.Captures("\"Hello\\\"Goodbye\"", result)

    [<TestMethod>]
    member _.``nrange consume`` () =
        let pattern = nrange "if" "end if"
        let mutable result = Result()

        result <- pattern.Consume("if\nif\nend if\nbacon\nend if\nfoobar")
        ResultAssert.Captures("if\nif\nend if\nbacon\nend if", result)

        result <- pattern.Consume("if\nif\nend if\nbacon\nfoobar")
        ResultAssert.Fails(result)